pallets/market/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+        Ok(asset_id)
+    }
+
+    fn apply_extension(
+        extension: &Extension<BlockNumberFor<T>, AssetIdOf<T>>,
+        who: &T::AccountId,
+        product_id: &AssetIdOf<T>,
+        amount: Balance,
+    ) -> DispatchResult {
+        match extension {
+            Extension::MaxAmount(max_amount) => {
+                ensure!(
+                    T::AssetInfoProvider::total_balance(product_id, who)?.saturating_add(amount)
+                        <= *max_amount,
+                    Error::<T>::MaxAmountExceeded
+                );
+            }
+            Extension::RequiredProduct(product_id, min_amount) => {
+                ensure!(
+                    T::AssetInfoProvider::total_balance(product_id, who)? >= *min_amount,
+                    Error::<T>::MissingRequiredProduct
+                );
+            }
+            Extension::DisallowedProduct(product_id) => {
+                ensure!(
+                    T::AssetInfoProvider::total_balance(product_id, who)? == 0,
+                    Error::<T>::HaveDisallowedProduct
+                );
+            }
+            Extension::Expirable(expiration) => {
+                let now = frame_system::Pallet::<T>::block_number();
+                Expirations::<T>::mutate(
+                    now.saturating_add(*expiration),
+                    (who, product_id),
+                    |amount_to_burn| {
+                        *amount_to_burn = amount_to_burn.saturating_add(amount);
+                    },
+                )
+            }
+        }
+        Ok(())
+    }
+
+    fn verify_extensions(
+        extensions: &[Extension<BlockNumberFor<T>, AssetIdOf<T>>],
+    ) -> DispatchResult {
+        let mut has_max_amount = false;
+        let mut has_expiration = false;
+        let mut required_products = BTreeSet::new();
+        let mut disallowed_products = BTreeSet::new();
+        for extension in extensions {
+            match extension {
+                Extension::MaxAmount(max_amount) => {
+                    ensure!(!has_max_amount, Error::<T>::MultipleExtensionsNotAllowed);
+                    has_max_amount = true;
+                    ensure!(!max_amount.is_zero(), Error::<T>::ZeroMaxAmount);
+                }
+                Extension::RequiredProduct(product_id, _min_amount) => {
+                    ensure!(
+                        !disallowed_products.contains(product_id),
+                        Error::<T>::AmbigiousProductRequirements
+                    );
+                    ensure!(
+                        required_products.insert(*product_id),
+                        Error::<T>::AmbigiousProductRequirements
+                    );
+                    ensure!(
+                        Products::<T>::contains_key(product_id),
+                        Error::<T>::ProductNotFound
+                    );
+                }
+                Extension::DisallowedProduct(product_id) => {
```

Comment:
I think we have to check that `product_id` in `RequiredProduct` & `DisallowedProduct` is exactly a product.
Because now it can be just a regular asset.
On the other hand, if we want to allow use all assets here (regular, NFT, etc.) we need to rename `...Product` into `...Asset`.